extension/dnsmasq/dnsmasq.js

```diff
@@ -1323,4 +1328,53 @@ module.exports = class DNSMASQ {
       log.error("Failed to clean up leftover config", err);
     }
   }
+
+  //save data in redis
+  //{mac:{ipv4Addr:ipv4Addr,name:name}}
+  //host: { ipv4Addr: '192.168.218.160',mac: 'F8:A2:D6:F1:16:53',name: 'LAPTOP-Lenovo' }
+  async setupLocalDeviceDomain(restart, hosts, isInit) {
+    const json = await rclient.getAsync(LOCAL_DEVICE_DOMAIN_KEY);
+    try {
+      let needUpdate = false;
+      const deviceDomainMap = JSON.parse(json) || {};
+      for (const host of hosts) {
+        //if user update device name and ip, should get them from redis not host
+        let hostName = await rclient.hgetAsync(hostTool.getMacKey(host.mac), "name");
+        let ipv4Addr = await rclient.hgetAsync(hostTool.getMacKey(host.mac), "ipv4Addr");
+        hostName = hostName ? hostName : hostTool.getHostname(host);
+        ipv4Addr = ipv4Addr ? ipv4Addr : host.ipv4Addr;
+        if (!deviceDomainMap[host.mac]) {
+          deviceDomainMap[host.mac] = {
+            ipv4Addr: ipv4Addr,
+            name: hostName
+          }
+          needUpdate = true;
+        } else {
+          const deviceDomain = deviceDomainMap[host.mac];
+          if ((deviceDomain.name != hostName || deviceDomain.ipv4Addr != ipv4Addr)) {
+            // need update
+            needUpdate = true;
+            deviceDomain.ipv4Addr = ipv4Addr;
+            deviceDomain.name = hostName;
+          }
+        }
+      }
+      if (needUpdate || isInit) {
+        await rclient.setAsync(LOCAL_DEVICE_DOMAIN_KEY, JSON.stringify(deviceDomainMap));
+        let localDeviceDomain = "";
+        for (const key in deviceDomainMap) {
+          const deviceDomain = deviceDomainMap[key]
+          //replace space to dot
+          const domain = deviceDomain.name.replace(/\s+/g, ".").toLowerCase();
```

Comment:
there may be other characters in the host name that may not be domain name friendly.  Like ","  "-" likely not allowed as first character. 

also deal with duplicated names?  host names can be duplicated

Reply:
How to deal with duplicated names?  override?

Reply:
Anything is fine:
1. override:  what ever we see last get that IP address ...  User will figure out and change the host name themselves.
2. add something unique to the name:  If we do this, we need to have UI present the name ... which is ugly, since this feature is for power users ... most 'normal' users won't care what a dot LAN is.
3. ignore both:  user will find out ... and change one of the names

I would go to 1, pick the criteria, like “latest" based on timestamp or something get the IP mapping. the other one ... don't. Absolutely avoid (2)